diff --git a/Source/ThirdParty/libwebrtc/Source/webrtc/test/fuzzers/sdp_integration_fuzzer.cc b/Source/ThirdParty/libwebrtc/Source/webrtc/test/fuzzers/sdp_integration_fuzzer.cc
index ece4b50505d8..e39e89f6c616 100644
--- a/Source/ThirdParty/libwebrtc/Source/webrtc/test/fuzzers/sdp_integration_fuzzer.cc
+++ b/Source/ThirdParty/libwebrtc/Source/webrtc/test/fuzzers/sdp_integration_fuzzer.cc
@@ -44,8 +44,10 @@ class FuzzerTest : public PeerConnectionIntegrationBaseTest {
       caller()->pc()->SetLocalDescription(sld_observer);
       EXPECT_TRUE_WAIT(sld_observer->called(), 100);
     }
+#if !defined(WEBRTC_WEBKIT_BUILD)
     // If there is an EXPECT failure, die here.
     RTC_CHECK(!HasFailure());
+#endif // !defined(WEBRTC_WEBKIT_BUILD)
   }
 
   // This test isn't using the test definition macros, so we have to
diff --git a/Source/ThirdParty/libwebrtc/Source/webrtc/test/gmock.h b/Source/ThirdParty/libwebrtc/Source/webrtc/test/gmock.h
index f137d080a4b3..2d4e26eb9738 100644
--- a/Source/ThirdParty/libwebrtc/Source/webrtc/test/gmock.h
+++ b/Source/ThirdParty/libwebrtc/Source/webrtc/test/gmock.h
@@ -14,7 +14,11 @@
 #include "rtc_base/ignore_wundef.h"
 
 RTC_PUSH_IGNORING_WUNDEF()
+#ifdef WEBRTC_WEBKIT_BUILD
+#include <gmock/gmock.h>
+#else
 #include "testing/gmock/include/gmock/gmock.h"
+#endif // WEBRTC_WEBKIT_BUILD
 RTC_POP_IGNORING_WUNDEF()
 
 #endif  // TEST_GMOCK_H_
diff --git a/Source/ThirdParty/libwebrtc/Source/webrtc/test/gtest.h b/Source/ThirdParty/libwebrtc/Source/webrtc/test/gtest.h
index fa4396420e4f..c4d9d1582e4c 100644
--- a/Source/ThirdParty/libwebrtc/Source/webrtc/test/gtest.h
+++ b/Source/ThirdParty/libwebrtc/Source/webrtc/test/gtest.h
@@ -14,8 +14,13 @@
 #include "rtc_base/ignore_wundef.h"
 
 RTC_PUSH_IGNORING_WUNDEF()
+#ifdef WEBRTC_WEBKIT_BUILD
+#include <gtest/gtest-spi.h>
+#include <gtest/gtest.h>
+#else
 #include "testing/gtest/include/gtest/gtest-spi.h"
 #include "testing/gtest/include/gtest/gtest.h"
+#endif // WEBRTC_WEBKIT_BUILD
 RTC_POP_IGNORING_WUNDEF()
 
 // GTEST_HAS_DEATH_TEST is set to 1 when death tests are supported, but appears
