diff -urN a/src/lib/ecore/ecore_main.c b/src/lib/ecore/ecore_main.c
--- a/src/lib/ecore/ecore_main.c	2014-03-10 19:45:48.000000000 +0900
+++ b/src/lib/ecore/ecore_main.c	2014-08-12 14:54:06.633710441 +0900
@@ -1584,8 +1584,6 @@
      {
 #ifndef _WIN32
         if (errno == EINTR) return -1;
-        else if (errno == EBADF)
-          _ecore_main_fd_handlers_bads_rem();
 #endif
      }
    if (ret > 0)
