diff --git a/gst/Makefile.am b/gst/Makefile.am
index d0e0062..d4bccea 100644
--- a/gst/Makefile.am
+++ b/gst/Makefile.am
@@ -13,6 +13,7 @@ AM_CFLAGS = \
     -Wall \
     -Wextra \
     -Werror \
+    -Wno-parentheses-equality \
     $(DEBUG_CFLAGS)
 
 lib_LTLIBRARIES = libopenwebrtc_gst.la
diff --git a/local/Makefile.am b/local/Makefile.am
index e2e7876..c6dc2f7 100644
--- a/local/Makefile.am
+++ b/local/Makefile.am
@@ -13,6 +13,7 @@ AM_CFLAGS = \
     -Wall \
     -Wextra \
     -Werror \
+    -Wno-parentheses-equality \
     $(DEBUG_CFLAGS)
 
 noinst_LTLIBRARIES = libopenwebrtc_local.la
diff --git a/owr/Makefile.am b/owr/Makefile.am
index 2af4735..dc3f1f2 100644
--- a/owr/Makefile.am
+++ b/owr/Makefile.am
@@ -14,6 +14,7 @@ AM_CFLAGS = \
     -Wall \
     -Wextra \
     -Werror \
+    -Wno-parentheses-equality \
     $(DEBUG_CFLAGS)
 
 lib_LTLIBRARIES = libopenwebrtc.la
diff --git a/transport/Makefile.am b/transport/Makefile.am
index 314ea53..5b5c749 100644
--- a/transport/Makefile.am
+++ b/transport/Makefile.am
@@ -15,6 +15,7 @@ AM_CFLAGS = \
     -Wall \
     -Wextra \
     -Werror \
+    -Wno-parentheses-equality \
     $(DEBUG_CFLAGS)
 
 noinst_LTLIBRARIES = libopenwebrtc_transport.la
