diff --git a/src/tint/lang/glsl/writer/BUILD.bazel b/src/tint/lang/glsl/writer/BUILD.bazel
index 390752a87..97fd1a439 100644
--- a/src/tint/lang/glsl/writer/BUILD.bazel
+++ b/src/tint/lang/glsl/writer/BUILD.bazel
@@ -63,14 +63,10 @@ cc_library(
     "//src/tint/utils/symbol",
     "//src/tint/utils/text",
     "//src/utils",
-  ] + select({
-    ":tint_build_glsl_writer": [
-      "//src/tint/lang/glsl/writer/common",
-      "//src/tint/lang/glsl/writer/printer",
-      "//src/tint/lang/glsl/writer/raise",
-    ],
-    "//conditions:default": [],
-  }),
+    "//src/tint/lang/glsl/writer/common",
+    "//src/tint/lang/glsl/writer/printer",
+    "//src/tint/lang/glsl/writer/raise",
+  ],
   copts = COPTS,
   visibility = ["//visibility:public"],
 )
diff --git a/src/tint/lang/glsl/writer/printer/BUILD.bazel b/src/tint/lang/glsl/writer/printer/BUILD.bazel
index e351c43ad..0e0aec5cb 100644
--- a/src/tint/lang/glsl/writer/printer/BUILD.bazel
+++ b/src/tint/lang/glsl/writer/printer/BUILD.bazel
@@ -66,12 +66,8 @@ cc_library(
     "//src/tint/utils/text",
     "//src/tint/utils/text_generator",
     "//src/utils",
-  ] + select({
-    ":tint_build_glsl_writer": [
-      "//src/tint/lang/glsl/writer/common",
-    ],
-    "//conditions:default": [],
-  }),
+    "//src/tint/lang/glsl/writer/common",
+  ],
   copts = COPTS,
   visibility = ["//visibility:public"],
 )
diff --git a/src/tint/lang/glsl/writer/raise/BUILD.bazel b/src/tint/lang/glsl/writer/raise/BUILD.bazel
index ef9dd4258..658ae65d5 100644
--- a/src/tint/lang/glsl/writer/raise/BUILD.bazel
+++ b/src/tint/lang/glsl/writer/raise/BUILD.bazel
@@ -80,12 +80,8 @@ cc_library(
     "//src/tint/utils/symbol",
     "//src/tint/utils/text",
     "//src/utils",
-  ] + select({
-    ":tint_build_glsl_writer": [
-      "//src/tint/lang/glsl/writer/common",
-    ],
-    "//conditions:default": [],
-  }),
+    "//src/tint/lang/glsl/writer/common",
+  ],
   copts = COPTS,
   visibility = ["//visibility:public"],
 )
