--- protobuf-3.0.0-beta-3/src/google/protobuf/arena.h   2016-05-16 11:34:04.000000000 -0700
+++ /usr/local/google/home/xyzzyz/chromium/src/third_party/protobuf/src/google/protobuf/arena.h 2016-05-17 14:38:50.109603998 -0700
@@ -213,7 +213,7 @@
 // well as RepeatedPtrField.
 
 #if __cplusplus >= 201103L
-class Arena final {
+class LIBPROTOBUF_EXPORT Arena final {
 #else
 class LIBPROTOBUF_EXPORT Arena {
 #endif
