diff --git a/sources/poky/meta/recipes-devtools/python/python3-markdown_3.4.1.bb b/sources/poky/meta/recipes-devtools/python/python3-markdown_3.4.1.bb
index e99c3310ac..b398cd8bc4 100644
--- a/sources/poky/meta/recipes-devtools/python/python3-markdown_3.4.1.bb
+++ b/sources/poky/meta/recipes-devtools/python/python3-markdown_3.4.1.bb
@@ -8,6 +8,6 @@ inherit pypi python_setuptools_build_meta
 PYPI_PACKAGE = "Markdown"
 SRC_URI[sha256sum] = "3b809086bb6efad416156e00a0da66fe47618a5d6918dd688f53f40c8e4cfeff"
 
-BBCLASSEXTEND = "native"
+BBCLASSEXTEND = "native nativesdk"
 
 RDEPENDS:${PN} += "${PYTHON_PN}-logging ${PYTHON_PN}-setuptools"
diff --git a/sources/poky/meta/recipes-devtools/python/python3-smartypants_2.0.0.bb b/sources/poky/meta/recipes-devtools/python/python3-smartypants_2.0.0.bb
index 05c94c390f..d089a89b95 100644
--- a/sources/poky/meta/recipes-devtools/python/python3-smartypants_2.0.0.bb
+++ b/sources/poky/meta/recipes-devtools/python/python3-smartypants_2.0.0.bb
@@ -9,6 +9,6 @@ PYPI_PACKAGE = "smartypants"
 SRC_URI += "file://0001-Change-hash-bang-to-python3.patch"
 SRC_URI[sha256sum] = "7812353a32022699a1aa8cd5626e01c94a946dcaeedaee2d0b382bae4c4cbf36"
 
-BBCLASSEXTEND = "native"
+BBCLASSEXTEND = "native nativesdk"
 
 UPSTREAM_CHECK_REGEX = "/${PYPI_PACKAGE}/(?P<pver>(?!2\.0\.1)(\d+[\.\-_]*)+)/"
diff --git a/sources/poky/meta/recipes-devtools/python/python3-typogrify_2.0.7.bb b/sources/poky/meta/recipes-devtools/python/python3-typogrify_2.0.7.bb
index 83e9b5eadb..826f126ad9 100644
--- a/sources/poky/meta/recipes-devtools/python/python3-typogrify_2.0.7.bb
+++ b/sources/poky/meta/recipes-devtools/python/python3-typogrify_2.0.7.bb
@@ -8,7 +8,7 @@ inherit pypi setuptools3
 PYPI_PACKAGE = "typogrify"
 SRC_URI[sha256sum] = "8be4668cda434163ce229d87ca273a11922cb1614cb359970b7dc96eed13cb38"
 
-BBCLASSEXTEND = "native"
+BBCLASSEXTEND = "native nativesdk"
 
 RDEPENDS:${PN} += "${PYTHON_PN}-smartypants"
 
diff --git a/sources/poky/meta/recipes-devtools/ruby/ruby_3.1.3.bb b/sources/poky/meta/recipes-devtools/ruby/ruby_3.1.3.bb
index c8454da3a9..85d6184a04 100644
--- a/sources/poky/meta/recipes-devtools/ruby/ruby_3.1.3.bb
+++ b/sources/poky/meta/recipes-devtools/ruby/ruby_3.1.3.bb
@@ -15,6 +15,7 @@ LIC_FILES_CHKSUM = "file://COPYING;md5=5b8c87559868796979806100db3f3805 \
 
 DEPENDS = "zlib openssl libyaml gdbm readline libffi"
 DEPENDS:append:class-target = " ruby-native"
+DEPENDS:append:class-nativesdk = " ruby-native"
 
 SHRT_VER = "${@oe.utils.trim_version("${PV}", 2)}"
 SRC_URI = "http://cache.ruby-lang.org/pub/ruby/${SHRT_VER}/ruby-${PV}.tar.gz \
@@ -139,4 +140,4 @@ FILES:${PN}-ptest:append:class-target = "\
     ${libdir}/ruby/${SHRT_VER}.0/*/-test- \
 "
 
-BBCLASSEXTEND = "native"
+BBCLASSEXTEND = "native nativesdk"
diff --git a/sources/poky/meta/recipes-devtools/unifdef/unifdef_2.12.bb b/sources/poky/meta/recipes-devtools/unifdef/unifdef_2.12.bb
index b42051b8b6..3e08b3a0a4 100644
--- a/sources/poky/meta/recipes-devtools/unifdef/unifdef_2.12.bb
+++ b/sources/poky/meta/recipes-devtools/unifdef/unifdef_2.12.bb
@@ -18,4 +18,4 @@ do_install() {
 	oe_runmake install DESTDIR=${D} prefix=${prefix}
 }
 
-BBCLASSEXTEND = "native"
+BBCLASSEXTEND = "native nativesdk"
diff --git a/sources/poky/meta/recipes-extended/gperf/gperf_3.1.bb b/sources/poky/meta/recipes-extended/gperf/gperf_3.1.bb
index c9f09c7931..4c32a5dc83 100644
--- a/sources/poky/meta/recipes-extended/gperf/gperf_3.1.bb
+++ b/sources/poky/meta/recipes-extended/gperf/gperf_3.1.bb
@@ -17,4 +17,4 @@ inherit autotools
 # it where to look explicitly (mirroring the behaviour of upstream's Makefile.devel).
 EXTRA_AUTORECONF += " -I ${S} --exclude=aclocal"
 
-BBCLASSEXTEND = "native"
+BBCLASSEXTEND = "native nativesdk"
diff --git a/sources/poky/meta/recipes-gnome/gi-docgen/gi-docgen_git.bb b/sources/poky/meta/recipes-gnome/gi-docgen/gi-docgen_git.bb
index 6a7124c3fe..000bc697a3 100644
--- a/sources/poky/meta/recipes-gnome/gi-docgen/gi-docgen_git.bb
+++ b/sources/poky/meta/recipes-gnome/gi-docgen/gi-docgen_git.bb
@@ -19,4 +19,4 @@ inherit setuptools3
 
 RDEPENDS:${PN} += "python3-asyncio python3-core python3-jinja2 python3-json python3-markdown python3-markupsafe python3-pygments python3-toml python3-typogrify python3-xml"
 
-BBCLASSEXTEND = "native"
+BBCLASSEXTEND = "native nativesdk"
diff --git a/sources/meta-raspberrypi/recipes-graphics/mesa/mesa-demos_%.bbappend b/sources/meta-raspberrypi/recipes-graphics/mesa/mesa-demos_%.bbappend
index abb11ec..ec0afa3 100644
--- a/sources/meta-raspberrypi/recipes-graphics/mesa/mesa-demos_%.bbappend
+++ b/sources/meta-raspberrypi/recipes-graphics/mesa/mesa-demos_%.bbappend
@@ -1,2 +1,2 @@
-# mesa-demos need libgles1 and userland driver does not have it
-COMPATIBLE_HOST:rpi = "${@bb.utils.contains('MACHINE_FEATURES', 'vc4graphics', '(.*)', 'null', d)}"
+# mesa-demos userland driver doesn't provide libgles1 and the EGL headers it provides break the mesa-demos build.
+PACKAGECONFIG:remove:rpi = "${@bb.utils.contains('MACHINE_FEATURES', 'vc4graphics', '', 'egl gles1', d)}"
