diff --git a/Source/ThirdParty/gmock/include/gmock/gmock-cardinalities.h b/Source/ThirdParty/gmock/include/gmock/gmock-cardinalities.h
index 533e604f326c..137327cbd10d 100644
--- a/Source/ThirdParty/gmock/include/gmock/gmock-cardinalities.h
+++ b/Source/ThirdParty/gmock/include/gmock/gmock-cardinalities.h
@@ -45,7 +45,7 @@
 #include <ostream>  // NOLINT
 
 #include "gmock/internal/gmock-port.h"
-#include "gtest/gtest.h"
+#include <gtest/gtest.h>
 
 GTEST_DISABLE_MSC_WARNINGS_PUSH_(4251 \
 /* class A needs to have dll-interface to be used by clients of class B */)
diff --git a/Source/ThirdParty/gmock/include/gmock/gmock-matchers.h b/Source/ThirdParty/gmock/include/gmock/gmock-matchers.h
index 0f67713776a6..383250c41370 100644
--- a/Source/ThirdParty/gmock/include/gmock/gmock-matchers.h
+++ b/Source/ThirdParty/gmock/include/gmock/gmock-matchers.h
@@ -274,7 +274,7 @@
 #include "gmock/internal/gmock-internal-utils.h"
 #include "gmock/internal/gmock-port.h"
 #include "gmock/internal/gmock-pp.h"
-#include "gtest/gtest.h"
+#include <gtest/gtest.h>
 
 // MSVC warning C5046 is new as of VS2017 version 15.8.
 #if defined(_MSC_VER) && _MSC_VER >= 1915
diff --git a/Source/ThirdParty/gmock/include/gmock/gmock-spec-builders.h b/Source/ThirdParty/gmock/include/gmock/gmock-spec-builders.h
index 78ca15d05e0f..5732ca41efae 100644
--- a/Source/ThirdParty/gmock/include/gmock/gmock-spec-builders.h
+++ b/Source/ThirdParty/gmock/include/gmock/gmock-spec-builders.h
@@ -78,7 +78,7 @@
 #include "gmock/gmock-matchers.h"
 #include "gmock/internal/gmock-internal-utils.h"
 #include "gmock/internal/gmock-port.h"
-#include "gtest/gtest.h"
+#include <gtest/gtest.h>
 
 #if GTEST_HAS_EXCEPTIONS
 #include <stdexcept>  // NOLINT
diff --git a/Source/ThirdParty/gmock/include/gmock/internal/gmock-internal-utils.h b/Source/ThirdParty/gmock/include/gmock/internal/gmock-internal-utils.h
index ead6d7c805e9..8cb9d1373375 100644
--- a/Source/ThirdParty/gmock/include/gmock/internal/gmock-internal-utils.h
+++ b/Source/ThirdParty/gmock/include/gmock/internal/gmock-internal-utils.h
@@ -47,7 +47,7 @@
 #include <vector>
 
 #include "gmock/internal/gmock-port.h"
-#include "gtest/gtest.h"
+#include <gtest/gtest.h>
 
 namespace testing {
 
diff --git a/Source/ThirdParty/gmock/include/gmock/internal/gmock-port.h b/Source/ThirdParty/gmock/include/gmock/internal/gmock-port.h
index 55ddfb6c6a3b..93fb7b65051b 100644
--- a/Source/ThirdParty/gmock/include/gmock/internal/gmock-port.h
+++ b/Source/ThirdParty/gmock/include/gmock/internal/gmock-port.h
@@ -54,7 +54,7 @@
 // here if it's truly specific to Google Mock.
 
 #include "gmock/internal/custom/gmock-port.h"
-#include "gtest/internal/gtest-port.h"
+#include <gtest/internal/gtest-port.h>
 
 #ifdef GTEST_HAS_ABSL
 #include "absl/flags/declare.h"
diff --git a/Source/ThirdParty/gmock/src/gmock-cardinalities.cc b/Source/ThirdParty/gmock/src/gmock-cardinalities.cc
index 92cde3484abf..680f2f09c29f 100644
--- a/Source/ThirdParty/gmock/src/gmock-cardinalities.cc
+++ b/Source/ThirdParty/gmock/src/gmock-cardinalities.cc
@@ -40,7 +40,7 @@
 #include <string>
 
 #include "gmock/internal/gmock-internal-utils.h"
-#include "gtest/gtest.h"
+#include <gtest/gtest.h>
 
 namespace testing {
 
diff --git a/Source/ThirdParty/gmock/src/gmock-internal-utils.cc b/Source/ThirdParty/gmock/src/gmock-internal-utils.cc
index 5c2ce0d57eb6..7df5eb69771e 100644
--- a/Source/ThirdParty/gmock/src/gmock-internal-utils.cc
+++ b/Source/ThirdParty/gmock/src/gmock-internal-utils.cc
@@ -48,7 +48,7 @@
 
 #include "gmock/gmock.h"
 #include "gmock/internal/gmock-port.h"
-#include "gtest/gtest.h"
+#include <gtest/gtest.h>
 
 namespace testing {
 namespace internal {
diff --git a/Source/ThirdParty/gmock/src/gmock-spec-builders.cc b/Source/ThirdParty/gmock/src/gmock-spec-builders.cc
index de8947163586..f5b75dd5925e 100644
--- a/Source/ThirdParty/gmock/src/gmock-spec-builders.cc
+++ b/Source/ThirdParty/gmock/src/gmock-spec-builders.cc
@@ -46,8 +46,8 @@
 #include <vector>
 
 #include "gmock/gmock.h"
-#include "gtest/gtest.h"
-#include "gtest/internal/gtest-port.h"
+#include <gtest/gtest.h>
+#include <gtest/internal/gtest-port.h>
 
 #if defined(GTEST_OS_CYGWIN) || defined(GTEST_OS_LINUX) || defined(GTEST_OS_MAC)
 #include <unistd.h>  // NOLINT
diff --git a/Source/ThirdParty/gmock/src/gmock_main.cc b/Source/ThirdParty/gmock/src/gmock_main.cc
index 6b5541271bfe..cb51cf89d732 100644
--- a/Source/ThirdParty/gmock/src/gmock_main.cc
+++ b/Source/ThirdParty/gmock/src/gmock_main.cc
@@ -30,7 +30,7 @@
 #include <iostream>
 
 #include "gmock/gmock.h"
-#include "gtest/gtest.h"
+#include <gtest/gtest.h>
 
 #if defined(GTEST_OS_ESP8266) || defined(GTEST_OS_ESP32) || \
     (defined(GTEST_OS_NRF52) && defined(ARDUINO))
