diff --git a/build/cmake/aom_configure.cmake b/build/cmake/aom_configure.cmake
index d79cb90366..4154e4dc9e 100644
--- a/build/cmake/aom_configure.cmake
+++ b/build/cmake/aom_configure.cmake
@@ -69,6 +69,9 @@ if(NOT AOM_TARGET_CPU)
     set(AOM_TARGET_CPU "x86")
   elseif(cpu_lowercase MATCHES "^arm" OR cpu_lowercase MATCHES "^mips")
     set(AOM_TARGET_CPU "${cpu_lowercase}")
+    if(CMAKE_SIZEOF_VOID_P EQUAL 4)
+      set(AOM_NEON_INTRIN_FLAG "-mfpu=neon")
+    endif()
   elseif(cpu_lowercase MATCHES "aarch64")
     set(AOM_TARGET_CPU "arm64")
   elseif(cpu_lowercase MATCHES "^ppc")
