From 475628c20e548a18c55e6bd8e9c3e3a73cf7192d Mon Sep 17 00:00:00 2001
From: Thibault Saunier <tsaunier@igalia.com>
Date: Fri, 31 May 2019 10:12:54 -0400
Subject: [PATCH] h264parse: Post a WARNING when data is broken

---
 gst/videoparsers/gsth264parse.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/gst/videoparsers/gsth264parse.c b/gst/videoparsers/gsth264parse.c
index 41a31eafa..fa64671de 100644
--- a/gst/videoparsers/gsth264parse.c
+++ b/gst/videoparsers/gsth264parse.c
@@ -1313,6 +1313,8 @@ gst_h264_parse_handle_frame (GstBaseParse * parse,
         /* broken nal at start -> arrange to skip it,
          * otherwise have it terminate current au
          * (and so it will be skipped on next frame round) */
+        GST_ELEMENT_WARNING (h264parse, STREAM, DECODE,
+            (NULL), ("Broken bit stream"));
         if (current_off == 0) {
           GST_DEBUG_OBJECT (h264parse, "skipping broken nal");
           *skipsize = nalu.offset;
-- 
2.21.0

