From 0ac86d7d0b8c2e4928c061c7cb93bf4b9b25529d Mon Sep 17 00:00:00 2001
From: Diego Pino Garcia <dpino@igalia.com>
Date: Thu, 17 Oct 2024 13:25:01 +0200
Subject: [PATCH] Add CMake flag to provision dependencies

---
 CMakeLists.txt | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 4c13d27f..4202dbf9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -50,6 +50,19 @@ if(CHECK_PIE_SUPPORTED)
   endif()
 endif()
 
+if(PROVISION_DEPENDENCIES)
+  # Run script to provision dependencies.
+  find_program (BASH_PROGRAM bash)
+  if(BASH_PROGRAM)
+    execute_process(
+      COMMAND ${BASH_PROGRAM} ${CMAKE_CURRENT_SOURCE_DIR}/deps.sh
+      RESULT_VARIABLE PROVISION_DEPENDENCIES_RESULT)
+  endif()
+  if(NOT PROVISION_DEPENDENCIES_RESULT EQUAL "0")
+    message(FATAL_ERROR "${CMAKE_CURRENT_SOURCE_DIR}/deps.sh failed with ${PROVISION_DEPENDENCIES_RESULT}")
+  endif()
+endif()
+
 ### Project build options:
 if(CXX_FUZZERS_SUPPORTED)
   # Enabled by default except on arm64, Windows and Apple builds.
-- 
2.46.0

