Index: icu-52~m1/source/config/mh-linux
===================================================================
--- icu-52~m1.orig/source/config/mh-linux	2013-09-14 18:53:23.284040467 -0400
+++ icu-52~m1/source/config/mh-linux	2013-09-14 18:53:23.284040467 -0400
@@ -21,7 +21,9 @@
 LD_RPATH_PRE = -Wl,-rpath,
 
 ## These are the library specific LDFLAGS
-LDFLAGSICUDT=-nodefaultlibs -nostdlib
+#LDFLAGSICUDT=-nodefaultlibs -nostdlib
+# Debian change: linking icudata as data only causes too many problems.
+LDFLAGSICUDT=
 
 ## Compiler switch to embed a library name
 # The initial tab in the next line is to prevent icu-config from reading it.
