From 2fda1d6acd43f6be9439eb4ae7a5abfabf415f3c Mon Sep 17 00:00:00 2001
From: Philippe Normand <philn@igalia.com>
Date: Mon, 5 Apr 2021 09:32:53 +0100
Subject: [PATCH] dav1d: Bump to version 0.8.2

---
 elements/components/dav1d.bst | 17 +++++++----------
 1 file changed, 7 insertions(+), 10 deletions(-)

diff --git a/elements/components/dav1d.bst b/elements/components/dav1d.bst
index 566dedd2..c679b340 100644
--- a/elements/components/dav1d.bst
+++ b/elements/components/dav1d.bst
@@ -1,15 +1,16 @@
 kind: meson
 
+build-depends:
+- public-stacks/buildsystem-meson.bst
+- components/nasm.bst
+
 depends:
-- filename: bootstrap-import.bst
-- filename: public-stacks/buildsystem-meson.bst
-  type: build
-- filename: components/nasm.bst
-  type: build
+- bootstrap-import.bst
 
 variables:
   meson-local: >-
     -Denable_tests=false
+    -Dxxhash_muxer=disabled
 
 public:
   bst:
@@ -22,8 +23,4 @@ sources:
 - kind: git_tag
   url: git_https:code.videolan.org/videolan/dav1d.git
   track: master
-  ref: 0.7.1-0-ge9df70c4348a3f9ba7269feacd17cfb57bf23852
-  match:
-  - '0.7.*'
-- kind: patch
-  path: patches/dav1d/nasm-syntax-fix.patch
+  ref: 0.8.2-0-gf06148e7c755098666b9c0ed97a672a51785413a
-- 
2.30.2

