From b53e9876f4611525b7a8f69a3110fbec51dcede0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Andreas=20M=C3=BCller?= <schnitzeltony@gmail.com>
Date: Thu, 19 Mar 2020 01:42:22 +0100
Subject: [PATCH] docs: Add man subdir only if manpage is enabled
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Andreas Müller <schnitzeltony@gmail.com>
---
 docs/meson.build | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/docs/meson.build b/docs/meson.build
index cc8d964..06db370 100644
--- a/docs/meson.build
+++ b/docs/meson.build
@@ -1,4 +1,6 @@
-subdir('man')
+if with_manpage
+  subdir('man')
+endif
 if with_gtkdoc
   subdir('reference/libsecret')
 endif
