--- gtk+-3.0.12-orig/gtk/gtkprinter.c  2011-07-27 18:34:32.000000000 -0700
+++ gtk+-3.0.12/gtk/gtkprinter.c       2012-03-15 04:43:10.260284244 -0700
@@ -1260,6 +1260,12 @@ gtk_enumerate_printers (GtkPrinterFunc f
         return;
     }
 
+  if (printer_list->backends == NULL)
+    {
+      free_printer_list (printer_list);
+      return;
+    }
+
   if (wait && printer_list->backends)
     {
       printer_list->loop = g_main_loop_new (NULL, FALSE);
