diff -ru llvm-3.5.0/lib/MC/MCObjectFileInfo.cpp llvm-3.5.0.patched/lib/MC/MCObjectFileInfo.cpp
--- llvm-3.5.0/lib/MC/MCObjectFileInfo.cpp	2014-06-25 14:41:52.000000000 +0200
+++ llvm-3.5.0.patched/lib/MC/MCObjectFileInfo.cpp	2014-11-02 11:58:39.410521253 +0100
@@ -274,6 +274,9 @@
   case Triple::mips64el:
     FDECFIEncoding = dwarf::DW_EH_PE_sdata8;
     break;
+  case Triple::x86_64:
+    FDECFIEncoding = dwarf::DW_EH_PE_pcrel;
+    break;
   default:
     FDECFIEncoding = dwarf::DW_EH_PE_pcrel | dwarf::DW_EH_PE_sdata4;
     break;
