From 6e5818deb54fdfa70f5798fc1df1d5c3eaba42df Mon Sep 17 00:00:00 2001
From: Bastien Nocera <hadess@hadess.net>
Date: Thu, 7 Jan 2016 15:48:47 +0100
Subject: Add *.xht as a glob for XHTML files


diff --git a/freedesktop.org.xml.in b/freedesktop.org.xml.in
index dc12655..48696d9 100644
--- a/freedesktop.org.xml.in
+++ b/freedesktop.org.xml.in
@@ -3669,6 +3669,7 @@ command to generate the output files.
     <sub-class-of type="application/xml"/>
     <generic-icon name="text-html"/>
     <glob pattern="*.xhtml"/>
+    <glob pattern="*.xht"/>
     <root-XML namespaceURI='http://www.w3.org/1999/xhtml' localName='html'/>
   </mime-type>
   <mime-type type="application/zip">
-- 
cgit v0.10.2

