diff --git a/third_party/pffft/src/pffft.h b/third_party/pffft/src/pffft.h
index 2bfa7b3ebcfb..bb6f78d4b795 100644
--- a/third_party/pffft/src/pffft.h
+++ b/third_party/pffft/src/pffft.h
@@ -83,6 +83,11 @@
 extern "C" {
 #endif
 
+#ifndef PFFFT_SIMD_DISABLE
+  // Detects compiler bugs with respect to simd instruction.
+  void validate_pffft_simd();
+#endif
+
   /* opaque struct holding internal stuff (precomputed twiddle factors)
      this struct can be shared by many threads as it contains only
      read-only data.  
